runtime/src/iree/hal/drivers/vulkan/debug_reporter.cc

```diff
@@ -56,6 +61,9 @@ iree_hal_vulkan_debug_utils_message_callback(
   }
   if (message_verbosity < reporter->min_verbosity) {
     fprintf(stderr, "[VULKAN] %c %s\n", severity_char, callback_data->pMessage);
+    if (reporter->check_errors) {
+      reporter->did_error = true;
```

Comment:
Cool, this caught a few issues on the CI (https://github.com/openxla/iree/actions/runs/7174549246/job/19536685446?pr=15884#step:7:1012): 
```
The following tests FAILED:
	117 - iree/tests/e2e/matmul/e2e_matmul_direct_i8_gpu_large_valhall_vulkan-spirv_vulkan (Subprocess aborted)
	215 - iree/tests/e2e/stablehlo_models/mnist_train_test/mnist_train_test_vulkan (Subprocess aborted)
	1526 - iree/hal/drivers/vulkan/cts/vulkan_buffer_mapping_test (Subprocess aborted)
	1527 - iree/hal/drivers/vulkan/cts/vulkan_command_buffer_test (Subprocess aborted)
	1535 - iree/hal/drivers/vulkan/cts/vulkan_pipeline_layout_test (Subprocess aborted)
```

```
[ RUN      ] CTS/command_buffer_test.FillBuffer_pattern1_size1_offset0_length1/vulkan
[VULKAN] ! Validation Error: [ VUID-VkMappedMemoryRange-size-01390 ] Object 0: handle = 0xe88693000000000c, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0xdd4e6d8b | vkFlushMappedMemoryRanges: Size in pMemRanges[0] is 0x4, which is not a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize (0x40) and offset + size (0x0 + 0x4 = 0x4) not equal to the memory size (0x40). The Vulkan spec states: If size is not equal to VK_WHOLE_SIZE, size must either be a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize, or offset plus size must equal the size of memory (https://vulkan.lunarg.com/doc/view/1.3.250.1/linux/1.3-extensions/vkspec.html#VUID-VkMappedMemoryRange-size-01390)
vulkan_command_buffer_test: /work/runtime/src/iree/hal/drivers/vulkan/debug_reporter.cc:155: void iree_hal_vulkan_debug_reporter_free(iree_hal_vulkan_debug_reporter_t *): Assertion `0' failed.
```

```
[ RUN      ] CTS/pipeline_layout_test.CreateWithTwoLayouts/vulkan
[VULKAN] ! Validation Error: [ VUID-VkPipelineLayoutCreateInfo-pSetLayouts-00293 ] Object 0: handle = 0x2b84e40, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7fc57626 | vkCreatePipelineLayout() Multiple push descriptor sets found. The Vulkan spec states: pSetLayouts must not contain more than one descriptor set layout that was created with VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR set (https://vulkan.lunarg.com/doc/view/1.3.250.1/linux/1.3-extensions/vkspec.html#VUID-VkPipelineLayoutCreateInfo-pSetLayouts-00293)
vulkan_pipeline_layout_test: /work/runtime/src/iree/hal/drivers/vulkan/debug_reporter.cc:155: void iree_hal_vulkan_debug_reporter_free(iree_hal_vulkan_debug_reporter_t *): Assertion `0' failed.
```

```
[ RUN      ] CTS/buffer_mapping_test.ZeroWithOffset/vulkan
[VULKAN] ! Validation Error: [ VUID-VkMappedMemoryRange-offset-00687 ] Object 0: handle = 0xe88693000000000c, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0x4f6e1263 | vkFlushMappedMemoryRanges: Offset in pMemRanges[0] is 0x4, which is not a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize (0x40). The Vulkan spec states: offset must be a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize (https://vulkan.lunarg.com/doc/view/1.3.250.1/linux/1.3-extensions/vkspec.html#VUID-VkMappedMemoryRange-offset-00687)
[VULKAN] ! Validation Error: [ VUID-VkMappedMemoryRange-size-01390 ] Object 0: handle = 0xe88693000000000c, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0xdd4e6d8b | vkFlushMappedMemoryRanges: Size in pMemRanges[0] is 0x8, which is not a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize (0x40) and offset + size (0x4 + 0x8 = 0xc) not equal to the memory size (0x10). The Vulkan spec states: If size is not equal to VK_WHOLE_SIZE, size must either be a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize, or offset plus size must equal the size of memory (https://vulkan.lunarg.com/doc/view/1.3.250.1/linux/1.3-extensions/vkspec.html#VUID-VkMappedMemoryRange-size-01390)
vulkan_buffer_mapping_test: /work/runtime/src/iree/hal/drivers/vulkan/debug_reporter.cc:155: void iree_hal_vulkan_debug_reporter_free(iree_hal_vulkan_debug_reporter_t *): Assertion `0' failed.
```

```
 21/362 Test  #117: iree/tests/e2e/matmul/e2e_matmul_direct_i8_gpu_large_valhall_vulkan-spirv_vulkan ....................Subprocess aborted***Exception:   1.57 sec
[VULKAN] ! Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-08740 ] Object 0: handle = 0x28cff90, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x6e224e9 | vkCreateShaderModule(): The SPIR-V Capability (DotProductKHR) was declared, but none of the requirements were met to use it. The Vulkan spec states: If pname:codeType is ename:VK_SHADER_CODE_TYPE_SPIRV_EXT, and pCode declares any of the capabilities listed in the SPIR-V Environment appendix, one of the corresponding requirements must be satisfied (https://vulkan.lunarg.com/doc/view/1.3.250.1/linux/1.3-extensions/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-08740)
```

Reply:
Iiiiinteresting... different errors on the NVIDIA CI: https://github.com/openxla/iree/actions/runs/7214784520/job/19657869428?pr=15884#step:7:1193 and the SwiftShader CI: https://github.com/openxla/iree/actions/runs/7214784520/job/19657869686?pr=15884#step:6:7897

Reply:
Looks like different device limits. Could make our code more robust to those...

I see `minStorageBufferOffsetAlignment                 = 0x00000010` on my machine and SwiftShader CI failed because `0x80` is not a multiple of `0x100`

Logs: https://github.com/openxla/iree/actions/runs/7214784520/job/19657869686?pr=15884#step:6:169
`[VULKAN] ! Validation Error: [ VUID-VkWriteDescriptorSet-descriptorType-00328 ] Object 0: handle = 0x25e4110, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xea08144e | vkUpdateDescriptorSets(): pDescriptorWrites[0].pBufferInfo[0].offset (0x80) must be a multiple of device limit minStorageBufferOffsetAlignment 0x100. The Vulkan spec states: If descriptorType is VK_DESCRIPTOR_TYPE_STORAGE_BUFFER or VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC, the offset member of each element of pBufferInfo must be a multiple of VkPhysicalDeviceLimits::minStorageBufferOffsetAlignment `

Reply:
Okay, got `vulkaninfo` output from SwiftShader on the CI too: https://github.com/openxla/iree/actions/runs/7216253462/job/19662332869?pr=15884#step:6:274
`minStorageBufferOffsetAlignment                 = 0x00000100`

So the "min storage buffer offset alignment" is
* NVIDIA GPUs: `0x10` (16 in decimal)
* SwiftShader: `0x100` (256 in decimal)

and we're using `0x80` (128 in decimal) as an offset.

Would be nice to see what our Android test devices report too.

---

I'd like to get us validation warnings clean before landing this change and removing SwiftShader.

Reply:
Hehe, found a TODO that relates to one of the errors: https://github.com/openxla/iree/blob/416e4b4ea42f86927947be45ae83388cf581a10f/runtime/src/iree/hal/drivers/vulkan/native_buffer.cc#L122-L130